## Podstawy programowania
### Praca domowa 03 [100 pkt]

Proszę użyć tego arkusza do rozwiązania zadań domowych. Na początek proszę zmienić 'nazwisko_imie' na Państwa nazwisko i imie, bez polskich znaków diakrytycznych.

Część zadań to powtórka z zajęć. Jak już Państwo zaprogramują wszystko, proszę zapisać arkusz (File->Save and Checkpoint) oraz zapisać arkusz na dysk (File->Download as->Notebook). Ten plik proszę wgrać na classroom. 

Jeszcze jedna uwaga: Używamy zmiennych oraz komentarzy (tam gdzie to w/g Państwa niezbędne).

In [ ]:
# współpraca: w znalezieniu solucji rozwiązania zadań 3 oraz 4 pomagał mi Igor Juraszek, resztę czerpałam z wiedzy nabytej na ćwiczeniach oraz internetu 
# przybliżony czas poświęcony na tą pracę domową: 4,5h

### Zadanie 1 [20 pkt] Sekwencje
Stwórz

* [3 pkt] listę pierwiastków kwadratowych liczb od 0.0001 do 3.1415 co 0.0001
* [3 pkt] krotkę liczb parzystych od -20 do 44
* [3 pkt] ciąg znaków odwrotności nieparzystych liczb naturalnych od 1 do 99
    $$s = \sum_{j = 1}^{99} \frac{1}{j}$$
* [3 pkt] zbiór (set) wszystkich liczb od -333 do 21345 podzielnych przez 8 i 11 
* [3 pkt] listę znaków wszystkich liczb całkowitych od -1000 do 0, w których znajdziemy cyfrę "3" (proszę zapoznać się/przypomnieć sobie o funkcji rzutującej `str()`)

Oblicz sumy liczb zmagazynowanych w powyższych zbiorach danych [5 $\times$ 1 pkt].

In [3]:
from math import sqrt

In [27]:
# lista pierwiastków kwadratowych liczb od 0.0001 do 3.1415 co 0.0001

lista = []
start = 0.0001
stop = 3.1415

while start <= stop:
    lista += [sqrt(start)]
    start += 0.0001
    
print(lista)

[0.01, 0.01414213562373095, 0.017320508075688773, 0.02, 0.022360679774997897, 0.024494897427831782, 0.026457513110645908, 0.028284271247461905, 0.030000000000000002, 0.0316227766016838, 0.033166247903554005, 0.03464101615137755, 0.0360555127546399, 0.03741657386773942, 0.038729833462074176, 0.04000000000000001, 0.04123105625617661, 0.04242640687119286, 0.04358898943540674, 0.0447213595499958, 0.045825756949558406, 0.0469041575982343, 0.04795831523312719, 0.04898979485566356, 0.049999999999999996, 0.05099019513592784, 0.05196152422706631, 0.0529150262212918, 0.05385164807134503, 0.0547722557505166, 0.05567764362830021, 0.05656854249492379, 0.05744562646538027, 0.05830951894845299, 0.05916079783099614, 0.059999999999999984, 0.06082762530298218, 0.061644140029689744, 0.06244997998398396, 0.06324555320336757, 0.06403124237432847, 0.06480740698407858, 0.06557438524301999, 0.06633249580710798, 0.06708203932499368, 0.06782329983125268, 0.06855654600401044, 0.06928203230275509, 0.0699999999999

In [30]:
# krotka liczb parzystych od -20 do 44

krotka = ()
start = -20
stop = 44

while start <= stop:
    if start % 2 == 0:
        krotka += (start, )
    start += 1

print(krotka)

(-20, -18, -16, -14, -12, -10, -8, -6, -4, -2, 0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44)


In [31]:
# ciąg znaków odwrotności nieparzystych liczb naturalnych od 1 do 99

ciag_znakow = ""
start = 1
stop = 99

while start <= 99:
    if start % 2 == 1:
        ciag_znakow += str(1 / start) + ", "
    start += 1
    
print(ciag_znakow)

1.0, 0.3333333333333333, 0.2, 0.14285714285714285, 0.1111111111111111, 0.09090909090909091, 0.07692307692307693, 0.06666666666666667, 0.058823529411764705, 0.05263157894736842, 0.047619047619047616, 0.043478260869565216, 0.04, 0.037037037037037035, 0.034482758620689655, 0.03225806451612903, 0.030303030303030304, 0.02857142857142857, 0.02702702702702703, 0.02564102564102564, 0.024390243902439025, 0.023255813953488372, 0.022222222222222223, 0.02127659574468085, 0.02040816326530612, 0.0196078431372549, 0.018867924528301886, 0.01818181818181818, 0.017543859649122806, 0.01694915254237288, 0.01639344262295082, 0.015873015873015872, 0.015384615384615385, 0.014925373134328358, 0.014492753623188406, 0.014084507042253521, 0.0136986301369863, 0.013333333333333334, 0.012987012987012988, 0.012658227848101266, 0.012345679012345678, 0.012048192771084338, 0.011764705882352941, 0.011494252873563218, 0.011235955056179775, 0.01098901098901099, 0.010752688172043012, 0.010526315789473684, 0.010309278350515

In [1]:
# zbiór (set) wszystkich liczb podzielnych przez 8 i 11 od -333 do 21345

zbior = set()
start = -333
stop = 21345

while start <= stop:
    if (start % 8 == 0) and (start % 11 == 0):
        zbior.update([start])
    start += 1
    
print(zbior) 

{0, 5632, 11264, 16896, 3080, 8712, 14344, 19976, 528, 6160, 11792, 17424, 3608, 9240, 14872, 20504, 1056, 6688, 12320, 17952, 4136, 9768, 15400, 21032, 1584, 7216, 12848, 18480, 4664, 10296, 15928, 19448, 2112, 7744, 13376, 19008, 5192, 10824, 16456, 2640, 8272, 13904, 19536, 88, 5720, 11352, 16984, 3168, 8800, 14432, 20064, 616, 6248, 11880, 17512, 3696, 9328, 14960, 20592, 1144, 6776, 12408, 18040, 4224, 9856, 15488, 21120, 1672, 7304, 12936, 18568, 4752, 10384, 16016, 2200, 7832, 13464, 19096, 5280, 10912, 16544, 2728, 8360, 13992, 19624, 176, 5808, 11440, 17072, 3256, 8888, 14520, 20152, 704, 6336, 11968, 17600, 3784, 9416, 15048, 20680, 1232, 6864, 12496, 18128, 4312, 8184, 9944, 15576, 21208, 1760, 7392, 13024, 18656, 4840, 10472, 16104, 2288, 7920, 13552, 19184, 5368, -264, 11000, 16632, 2816, 8448, 14080, 19712, 264, 5896, 11528, 17160, 3344, 8976, 14608, 20240, 792, 6424, 12056, 17688, 3872, 9504, 15136, 20768, 1320, 6952, 12584, 18216, 4400, 10032, 15664, 21296, 1848, 7480, 

In [4]:
# lista wszystkich liczb całkowitych od -1000 do 0, w których znajdziemy cyfrę "3"

lista = []
start = -1000

while start <= 0:
    if (str(start).count("3")):
        lista += [start]
    start += 1

print(lista)

[-993, -983, -973, -963, -953, -943, -939, -938, -937, -936, -935, -934, -933, -932, -931, -930, -923, -913, -903, -893, -883, -873, -863, -853, -843, -839, -838, -837, -836, -835, -834, -833, -832, -831, -830, -823, -813, -803, -793, -783, -773, -763, -753, -743, -739, -738, -737, -736, -735, -734, -733, -732, -731, -730, -723, -713, -703, -693, -683, -673, -663, -653, -643, -639, -638, -637, -636, -635, -634, -633, -632, -631, -630, -623, -613, -603, -593, -583, -573, -563, -553, -543, -539, -538, -537, -536, -535, -534, -533, -532, -531, -530, -523, -513, -503, -493, -483, -473, -463, -453, -443, -439, -438, -437, -436, -435, -434, -433, -432, -431, -430, -423, -413, -403, -399, -398, -397, -396, -395, -394, -393, -392, -391, -390, -389, -388, -387, -386, -385, -384, -383, -382, -381, -380, -379, -378, -377, -376, -375, -374, -373, -372, -371, -370, -369, -368, -367, -366, -365, -364, -363, -362, -361, -360, -359, -358, -357, -356, -355, -354, -353, -352, -351, -350, -349, -348, -34

### Zadanie 2 Znów liczba urodzenia [20 pkt]
Za pomocą pętli (nie ręcznie) oblicz swoją [liczbę urodzenia](https://pl.wikipedia.org/wiki/Numerologia).

In [32]:
# rozwiązanie tutaj

data = input(str("Twoja data urodzenia (dd-mm-rrrr): "))

start = 0
suma = 0

while start < len(data):
    if data[start].isdigit():
        suma += int(data[start])
    start += 1

while suma >= 10:
    data = str(suma)
    suma = int(data[0]) + int(data[1])
else:
    print(suma)

Twoja data urodzenia (dd-mm-rrrr): 20-06-2001
2


### Zadanie 3 Dobre hasło [20 pkt]
Większość serwisów oferuje obecnie tworzenie kont, do których to trzeba podać relatywnie silnie hasło, takie na które składa się 
* przynajmniej 8 znaków, w tym:
* przynajmniej 1 duża litera
* przynajmniej 1 cyfra
* przynajmniej 1 znak specjalny (`!@#%^&*()_+}{;:.,/?`)

Napisz program, który sprawdza czy hasło podane w zmiennej `haslo` spełnia powyższe założenia. Jeżeli spełnia to proszę poprawnie sformatować informację przekazaną funkcją `print`

    'haslo <hedh7@jA> jest poprawne'
    
lub

    'haslo <hedh7@ja> jest niepoprawne'

In [40]:
# rozwiązanie tutaj

In [45]:
haslo = input(str("Wpisz przykładowe hasło: "))

duze_litery = 0
cyfry = 0
znaki_specjalne = 0
tablica_znakow = ['!', '@', '#', '%', '^', '&', '*', '(', ')', '_', '+', '}', '{', ';', ':', '.', ',', '/', '?']

for znak in haslo:
    if znak.isupper():
        duze_litery += 1
    if znak.isdigit():
        cyfry += 1
    if znak in tablica_znakow:
        znaki_specjalne += 1
if (len(haslo) >= 8) and (duze_litery >= 1) and (cyfry >= 1) and (znaki_specjalne >= 1):
    print(f'haslo <{haslo}> jest poprawne')
else:
    print(f'haslo <{haslo}> jest niepoprawne')

Wpisz przykładowe hasło: hedh7@ja
haslo <hedh7@ja> jest niepoprawne


### Wstęp do Zadania 4

Za pomocą biblioteki `random` możemy tworzyć liczby pseudo-losowe. Można też losowo wybierać elementy sekwencji. Więcej informacji [np. tutaj](https://pynative.com/python-random-module/).

In [ ]:
import string
import random
liczba_01 = random.random()
liczba_10_99 = random.randint(10, 99)
litery = random.sample(string.ascii_letters, 5)

print(f'liczba z przedziału [0, 1): {liczba_01}') 
print(f'liczba z przedziału [10, 99): {liczba_10_99}')
print(f'5 losowych liter z alfabetu: {litery}')

### Zadanie 4 Generuj dobre hasło [20 pkt]
Wygeneruj hasło spełniające założenia z zadania 3.

In [75]:
# rozwiązanie tutaj

import string
import random

haslo = ""

znaki_hasla = []

tablica_znakow = ['!', '@', '#', '%', '^', '&', '*', '(', ')', '_', '+', '}', '{', ';', ':', '.', ',', '/', '?']

# do listy zawierającej znaki hasła dodajemy kolejno 1 losową cyfrę, 
# 1 dużą literę alfabetu, 5 dowolnych liter oraz 1 znak specjalny z tablicy znaków

znaki_hasla += str(random.randint(0, 9))
znaki_hasla += random.sample(string.ascii_letters.upper(), 1)
znaki_hasla += random.sample(string.ascii_letters, 5)
znaki_hasla += random.sample(tablica_znakow, 1)

# elementy hasła dowolnie się mieszają

random.shuffle(znaki_hasla)

# za pomocą pętli for wszystkie elementy tablicy zapisujemy do ciągu znaków

for znak in znaki_hasla:
    haslo += znak
print(f"Twoje hasło to: {haslo}")

Twoje hasło to: t9pRk;ZI


### Zadanie 5 [20 pkt] 1089
Proszę napisać program który dla konkretnej trzycyfrowej `ABC` (np: `543`) liczby oblicza po kolei

1. odwraca cyfry liczby oryginalnej `CBA` (`345`)

2. odejmuje liczbę odwróconą od oryginalnej `XYZ = ABC - CBA (543 - 345 = 198)`

3. odwracamy cyfry obliczonej różnicy `ZYX (891)`

4. dodajemy do siebie różnicę (`XYZ`) oraz jej odwrócenie (`ZYX`): 
`wynik = XYZ + ZYX  (198 + 891 = 1089)`

5. drukujemy powyższy `wynik`

Proszę posprawdzać taki algorytm dla kilku liczb. Jeżeli cyfry w oryginalnej liczbie ABC będą ułożone malejąco (jak w powyższym przykładzie), to wynik obliczeń będzie wynosił zawsze `1089`.

In [ ]:
# rozwiązanie tutaj

In [112]:
# z treści zadania zrozumiałam, iż cyfry zawarte w liczbie trzycyfrowej muszą być od siebie różne,
# to znaczy ABC, a nie ABA, CCA itp.

liczba_trzycyfrowa = input(str("Wpisz liczbę trzycyfrową: "))

while len(liczba_trzycyfrowa) != 3:
    liczba_trzycyfrowa = input(str("BŁĄD! Wpisz liczbę trzycyfrową: "))
    
print(f"Wpisana liczba trzycyfrowa: {liczba_trzycyfrowa}")

# używam metod dla stringów 

odwrocona_liczba = str(liczba_trzycyfrowa)[::-1]
print(f"Podana liczba po jej odwróceniu: {odwrocona_liczba}")

roznica = abs(int(liczba_trzycyfrowa) - int(odwrocona_liczba))
print(f"Różnica podanej liczby trzycyfrowej oraz jej odwrócenia: {roznica}")

odwrocona_roznica = str(roznica)[::-1]
print(f"Odwrócona różnica: {odwrocona_roznica}")

wynik = int(odwrocona_roznica) + int(roznica)
print(f"Wynik: {wynik}")

Wpisz liczbę trzycyfrową: 456
Wpisana liczba trzycyfrowa: 456
Podana liczba po jej odwróceniu: 654
Różnica podanej liczby trzycyfrowej oraz jej odwrócenia: 198
Odwrócona różnica: 891
Wynik: 1089


In [111]:
# przystosowuję program do obliczenia naraz kilku konkretnych liczb z listy, by sprawdzić jego działanie

liczby = [456, 897, 123, 943, 489]
for liczba in liczby:
    liczba_trzycyfrowa = str(liczba)

    print(f"Wpisana liczba trzycyfrowa: {liczba_trzycyfrowa}")

    odwrocona_liczba = str(liczba_trzycyfrowa)[::-1]
    print(f"Podana liczba po jej odwróceniu: {odwrocona_liczba}")

    roznica = abs(int(liczba_trzycyfrowa) - int(odwrocona_liczba))
    print(f"Różnica podanej liczby trzycyfrowej oraz jej odwrócenia: {roznica}")

    odwrocona_roznica = str(roznica)[::-1]
    print(f"Odwrócona różnica: {odwrocona_roznica}")

    wynik = int(odwrocona_roznica) + int(roznica)
    print(f"Wynik: {wynik} \n")

Wpisana liczba trzycyfrowa: 456
Podana liczba po jej odwróceniu: 654
Różnica podanej liczby trzycyfrowej oraz jej odwrócenia: 198
Odwrócona różnica: 891
Wynik: 1089 

Wpisana liczba trzycyfrowa: 897
Podana liczba po jej odwróceniu: 798
Różnica podanej liczby trzycyfrowej oraz jej odwrócenia: 99
Odwrócona różnica: 99
Wynik: 198 

Wpisana liczba trzycyfrowa: 123
Podana liczba po jej odwróceniu: 321
Różnica podanej liczby trzycyfrowej oraz jej odwrócenia: 198
Odwrócona różnica: 891
Wynik: 1089 

Wpisana liczba trzycyfrowa: 943
Podana liczba po jej odwróceniu: 349
Różnica podanej liczby trzycyfrowej oraz jej odwrócenia: 594
Odwrócona różnica: 495
Wynik: 1089 

Wpisana liczba trzycyfrowa: 489
Podana liczba po jej odwróceniu: 984
Różnica podanej liczby trzycyfrowej oraz jej odwrócenia: 495
Odwrócona różnica: 594
Wynik: 1089 

